streamer/controller_node.py

```diff
@@ -196,44 +243,55 @@ def start(self, output_dir: str,
 
-    self._nodes.append(TranscoderNode(input_config,
-                                      pipeline_config,
+    self._nodes.append(TranscoderNode(inputs,
+                                      self._pipeline_config,
                                       outputs))
 
-    self._nodes.append(PackagerNode(pipeline_config,
+    output_dir = self._output_dir
+    
+    # If the inputs list was a period in multiperiod_inputs_list, create a nested directory
+    # and put that period in it.
+    if period_dir:
+      output_dir = os.path.join(output_dir, period_dir)
+      os.mkdir(output_dir)
+
+    self._nodes.append(PackagerNode(self._pipeline_config,
                                     output_dir,
                                     outputs))
 
-    if bucket_url:
-      cloud_temp_dir = os.path.join(self._temp_dir, 'cloud')
-      os.mkdir(cloud_temp_dir)
-
-      self._nodes.append(CloudNode(output_dir,
-                                   bucket_url,
-                                   cloud_temp_dir,
-                                   self.is_vod()))
-
-    for node in self._nodes:
-      node.start()
-    return self
-
   def check_status(self) -> ProcessStatus:
     """Checks the status of all the nodes.
 
-    If one node is errored, this returns Errored; otherwise if one node is
-    finished, this returns Finished; this only returns Running if all nodes are
-    running.  If there are no nodes, this returns Finished.
+    If one node is errored, this returns ERRORED; otherwise if one Transcoder, Packager or PeriodConcat 
+    node is running, this returns RUNNING; this only returns FINISHED if all Transcoder, Packager and
+    PeriodConcat nodes are finished. If there are no nodes, this returns FINISHED.
     """
-    if not self._nodes:
-      return ProcessStatus.Finished
+    
+    values: Set[int] = set([ProcessStatus.FINISHED.value])
 
-    value = max(node.check_status().value for node in self._nodes)
+    for node in self._nodes:
+      status: ProcessStatus = node.check_status()
+      if isinstance(node, (TranscoderNode, PackagerNode, PeriodConcatNode)):
+        values.add(status.value)
+      # Return ERRORED even if the node wasn't an instance of the checked nodes.
+      if status == ProcessStatus.ERRORED:
+        return ProcessStatus.ERRORED
+        
+    value: int = max(values)
     return ProcessStatus(value)
 
   def stop(self) -> None:
-    """Stop all nodes."""
-    status = self.check_status()
-    for node in self._nodes:
+    """Stop all nodes.
+    Each Transcoder and Packager nodes are paired together, the status sent to each one of them
+    in the stop(status) is the status of the other.
+    """
+    
+    for i, node in enumerate(self._nodes):
+      status: Optional[ProcessStatus] = None
+      if isinstance(node, TranscoderNode):
+        status = self._nodes[i+1].check_status()
+      elif isinstance(node, PackagerNode):
+        status = self._nodes[i-1].check_status()
```

Comment:
It's no more the case that there is only one transcoder and packager node, i think putting each type in a separate list will be less confusing, right?

Reply:
I don't think this change is necessary.  stop() is only called on the controller when we're stopping everything.  There are only two cases for this:

1. check_status() on controller is Finished, meaning everything is finished, so the pairs don't matter because all status are the same
2. check_status() on controller is Running or Errored, meaning something failed or we shut down early, so the pairs don't matter because we want to force-kill anything still running